File: src/session.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class TokenUsage:
    input: int = 0
    output: int = 0
    cache_read: int = 0
    cache_create: int = 0
    turns: int = 0


@dataclass
class Session:
    source: str
    session_id: str
    path: str
    cwd: str | None = None
    first_ts: str | None = None
    last_ts: str | None = None
    title: str | None = None
    subtitle: str | None = None
    first_prompt: str | None = None
    last_prompt: str | None = None
    prompts: list[tuple[str, str]] = field(default_factory=list)
    asst_count: int = 0
    tokens: TokenUsage | None = None
    resume_argv: list[str] = field(default_factory=list)


@dataclass
class Filters:
    days: int | None = 3
    date: str | None = None
    all_time: bool = False
    project: str | None = None
    limit: int | None = None
    source: str | None = None

File: internal/session/session.go
// Package session defines the provider-agnostic session schema shared by
// providers, renderers, the TUI, and the exec dispatcher.
package session

// TokenUsage aggregates assistant-turn usage across a session.
type TokenUsage struct {
	Input       int64 `json:"input"`
	Output      int64 `json:"output"`
	CacheRead   int64 `json:"cache_read"`
	CacheCreate int64 `json:"cache_create"`
	Turns       int64 `json:"turns"`
}

// Prompt is one user prompt with its timestamp (empty string when unknown).
type Prompt struct {
	TS   string
	Text string
}

// Session is the common schema. String fields use "" for "absent" (the
// Python implementation used None); render/json restores explicit nulls.
type Session struct {
	Source       string
	SessionID    string
	Path         string
	ProjectLabel string
	Cwd          string
	FirstTS      string
	LastTS       string
	Title        string
	Subtitle     string
	FirstPrompt  string
	LastPrompt   string
	Prompts      []Prompt
	AsstCount    int
	Tokens       *TokenUsage // nil when no assistant turns carried usage
	ResumeArgv   []string
}

// Filters mirrors the CLI surface. Days < 0 means "unset" (provider default).
type Filters struct {
	Days    int
	Date    string
	AllTime bool
	Project string
	Limit   int
	Source  string
}
